api/envoy/service/network_ext_proc/v3/network_external_processor.proto

```diff
@@ -0,0 +1,155 @@
+  // they are received to maintain proper sequencing of network traffic.
+  rpc Process(stream ProcessingRequest) returns (stream ProcessingResponse) {
+  }
+}
+
+// The payload data from network layer
+message Data {
+  // The raw payload data
+  bytes data = 1;
+
+  // Indicates whether this is the last data frame in the current direction.
+  // The external processor should still respond to this message even
+  // if there is no more data expected in this direction.
+  bool end_of_stream = 2;
+}
+
+// ProcessingRequest contains data sent from Envoy to the external processing server.
+// Each request contains either read data (from client) or write data (to client)
+// along with optional metadata.
+message ProcessingRequest {
+  // ReadData contains the network data intercepted in the request path (client to server).
+  // This is sent to the external processor when data arrives from the downstream client.
+  // If this is set, write_data should not be set.
+  Data read_data = 1;
+
+  // WriteData contains the network data intercepted in the response path (server to client).
+  // This is sent to the external processor when data arrives from the upstream server.
+  // If this is set, read_data should not be set.
+  Data write_data = 2;
+
+  // Optional metadata associated with the request.
+  // This can include connection properties, filter configuration, and any other
+  // contextual information that might be useful for processing decisions.
+  //
+  // The metadata is not automatically propagated from request to response.
+  // The external processor must include any needed metadata in its response.
+  config.core.v3.Metadata metadata = 3;
+}
+
+// ProcessingResponse contains the response from the external processing server to Envoy.
+// Each response corresponds to a ProcessingRequest and indicates how the network
+// traffic should be handled.
+// [#next-free-field: 6]
+message ProcessingResponse {
+  // DataProcessedStatus indicates whether the data was modified by the external processor.
+  enum DataProcessedStatus {
+    // The data remains unchanged. Envoy will use the original data.
+    // This is useful when the processor only wants to inspect but not modify the traffic.
+    UNMODIFIED = 0;
+
+    // The data has been modified. Envoy will use the modified data provided in the response.
+    // This allows the processor to transform the network traffic as needed.
+    MODIFIED = 1;
+  }
+
+  // ConnectionStatus determines what happens to the connection after processing.
+  enum ConnectionStatus {
+    // Continue normal processing of the connection.
+    // This is the default behavior and allows traffic to flow normally.
+    CONTINUE = 0;
+
+    // Close the connection gracefully (e.g., with a proper TCP FIN).
+    // This allows for clean connection termination with a normal close sequence.
+    // Any buffered data will be sent before closing.
+    CLOSE = 1;
```

Comment:
This should probably be HALF_CLOSE. I do not think you can close TCP gracefully in any other way.